File: app-ios/Packages/module-ai/Sources/ModuleAI/Application/UseCases/History/LoadHistoryPageUseCase.swift
//
//  LoadHistoryPageUseCase.swift
//  ModuleAI
//
//  Created by Codex on 2026/1/8.
//

import Foundation

struct LoadHistoryPageUseCase {
    private let repository: HistoryRepository
    private let dateFormatter: DateFormatter

    init(repository: HistoryRepository, dateFormatter: DateFormatter) {
        self.repository = repository
        self.dateFormatter = dateFormatter
    }

    func execute(
        pageNo: Int,
        pageSize: Int,
        existingGroups: [HistoryRecordGroup]
    ) async throws -> HistoryStateSnapshot {
        let pageModel = try await repository.loadHistoryPage(pageNo: pageNo, pageSize: pageSize)
        let groupedNewRecords = HistoryApplicationMapper.groupRecords(
            pageModel.records,
            dateFormatter: dateFormatter
        )

        let mergedGroups: [HistoryRecordGroup]
        if (pageModel.currentPage ?? 1) == 1 || existingGroups.isEmpty {
            mergedGroups = groupedNewRecords
        } else {
            var merged = existingGroups
            HistoryApplicationMapper.mergeGroups(existing: &merged, new: groupedNewRecords)
            mergedGroups = merged
        }

        return HistoryStateSnapshot(
            pageModel: pageModel,
            recordGroups: mergedGroups
        )
    }
}

File: app-android/feature/history/src/main/java/com/aidatainsight/android/feature/history/application/usecase/LoadHistoryPageUseCase.kt
package com.aidatainsight.android.feature.history.application.usecase

import com.aidatainsight.android.feature.history.application.HistoryApplicationMapper
import com.aidatainsight.android.feature.history.application.model.HistoryRecordGroup
import com.aidatainsight.android.feature.history.application.model.HistoryStateSnapshot
import com.aidatainsight.android.feature.history.domain.HistoryRepository

class LoadHistoryPageUseCase(
    private val repository: HistoryRepository,
) {
    suspend operator fun invoke(
        currentPage: Int,
        pageSize: Int,
        existingGroups: List<HistoryRecordGroup>,
    ): HistoryStateSnapshot {
        val page = repository.loadHistoryPage(currentPage = currentPage, pageSize = pageSize)
        val newGroups = HistoryApplicationMapper.groupRecords(page.records)
        val groups = if ((page.currentPage ?: currentPage) == 1 || existingGroups.isEmpty()) {
            newGroups
        } else {
            HistoryApplicationMapper.mergeGroups(existingGroups, newGroups)
        }
        return HistoryStateSnapshot(page = page, groups = groups)
    }
}
